Remove 4267 suppression from mono

am11 opened the pull request:
Contributes to #66154

cc @AaronRobinsonMSFT, @lambdageek

am11 committed:
Fix a mono warning on linux arm64

am11 committed:
Remove 4267 suppression from mono

am11 committed:
Remove trailing whitespaces from changeset

am11 commented:
AOT failure is unrelated, opened #66556.

AaronRobinsonMSFT commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -4332,7 +4332,7 @@ mono_profiler_fire_buffered_gc_event_root_register (
 	uintptr_t root_key = (uintptr_t)key;
 	const char *root_name = (name ? name : "");
-	uint32_t root_name_len = strlen (root_name) + 1;
+	uint32_t root_name_len = (uint32_t)strlen (root_name) + 1;
```
```suggestion
	size_t root_name_len = strlen (root_name) + 1;
```

AaronRobinsonMSFT commented on src/mono/mono/metadata/appdomain.c:
```diff
@@ -643,7 +643,7 @@ real_load (gchar **search_path, const gchar *culture, const gchar *name, const M
 
 	filename =  g_strconcat (name, ".dll", (const char*)NULL);
-	len = strlen (filename);
+	len = (gint)strlen (filename);
```
I'd update `len` to be `size_t` instead.

AaronRobinsonMSFT commented on src/mono/mono/metadata/class.c:
```diff
@@ -3285,7 +3285,7 @@ mono_class_from_name_checked_aux (MonoImage *image, const char* name_space, cons
 	if ((nested = (char*)strchr (name, '/'))) {
 		int pos = nested - name;
-		int len = strlen (name);
+		int len = (int)strlen (name);
```
```suggestion
		size_t len = strlen (name);
```

AaronRobinsonMSFT commented on src/mono/mono/metadata/debug-helpers.c:
```diff
@@ -703,7 +703,7 @@ dis_one (GString *str, MonoDisHelper *dh, MonoMethod *method, const unsigned cha
 			/* The blob might not be 2 byte aligned */
 			blob2 = g_malloc ((len2 * 2) + 1);
-			memcpy (blob2, blob, len2 * 2);
+			memcpy (blob2, blob, (int)len2 * 2);
```
This is odd to me. I thought `memcpy`'s last arg is `size_t`.

AaronRobinsonMSFT commented on src/mono/mono/metadata/domain.c:
```diff
@@ -828,7 +828,7 @@ get_runtime_by_version (const char *version)
 	}
 
-	vlen = strlen (version);
+	vlen = (int)strlen (version);
```
Update `int vlen;` to `size_t vlen;`

AaronRobinsonMSFT commented on src/mono/mono/mini/dwarfwriter.c:
```diff
@@ -204,7 +204,7 @@ emit_escaped_string (MonoDwarfWriter *w, char *value)
 	int i, len;
 
-	len = strlen (value);
+	len = (int)strlen (value);
```
This pattern seems very common and it concerns me we are going to have to update this for other CA warnings. Many of them make sense but it seems like we are sort of hiding issues.

AaronRobinsonMSFT commented on src/mono/mono/mini/exceptions-amd64.c:
```diff
@@ -1413,7 +1413,7 @@ mono_arch_unwindinfo_insert_range_in_table (const gpointer code_block, gsize blo
 			new_entry->begin_range = begin_range;
 			new_entry->end_range = end_range;
-			new_entry->rt_funcs_max_count = (block_size / MONO_UNWIND_INFO_RT_FUNC_SIZE) + 1;
+			new_entry->rt_funcs_max_count = (DWORD)((block_size / MONO_UNWIND_INFO_RT_FUNC_SIZE) + 1);
```
Instead of `DWORD` can you try `uint32_t` instead?

am11 replied:
I used the types (aliases), which compiler was expecting in a given context. Generally, mono uses `g<type>` types. Note that compiler warns with the same type name which the given expression is expected to have (e.g. which the function prototype has)

What do you think, shall we keep it?

AaronRobinsonMSFT replied:
No, you're right here. Let's keep it unless the Mono folks push back.

lateralusX replied:
This is Windows specific code and the underlying struct is using DWORD for this field since it is used with Win32 functions assuming DWORD's, so I think the cast is fine.

AaronRobinsonMSFT commented on src/mono/mono/mini/mini-generic-sharing.c:
```diff
@@ -4038,7 +4038,7 @@ get_shared_gparam_name (MonoTypeEnum constraint, const char *name)
 		t.type = constraint;
 		tname = mono_type_full_name (&t);
-		int len = strlen (tname);
+		int len = (int)strlen (tname);
```
```suggestion
		size_t len = strlen (tname);
```

AaronRobinsonMSFT commented on src/mono/mono/utils/mono-dl.c:
```diff
@@ -469,7 +469,7 @@ dl_build_path (const char *directory, const char *name, void **iter, dl_library_
 
 	if (need_prefix) {
-		prlen = strlen (mono_dl_get_so_prefix ());
+		prlen = (int)strlen (mono_dl_get_so_prefix ());
```
This seems like a good example where `prlen` as `size_t` is the right thing and it should just work. No?

AaronRobinsonMSFT commented on src/mono/mono/utils/mono-stdlib.c:
```diff
@@ -34,7 +34,7 @@ mono_mkstemp (char *templ)
 	int len;
 
-	len = strlen (templ);
+	len = (int)strlen (templ);
```
`size_t` instead?

AaronRobinsonMSFT commented on src/mono/mono/utils/monobitset.h:
```diff
@@ -50,7 +50,7 @@ enum {
     MonoBitSet *tmp_src = (src); \
     MonoBitSet *tmp_dest = (dest); \
     int i, size; \
```
```suggestion
    size_t i, size; \
```

Then the cast can be removed.

AaronRobinsonMSFT commented on src/mono/mono/utils/monobitset.h:
```diff
@@ -59,7 +59,7 @@ enum {
     MonoBitSet *tmp_src = (src); \
     MonoBitSet *tmp_dest = (dest); \
     int i, size; \
```
```suggestion
    size_t i, size; \
```
Then the cast can be removed.

AaronRobinsonMSFT commented on src/mono/mono/utils/parse.c:
```diff
@@ -29,7 +29,7 @@ gboolean
 {
 	char *endptr;
-	int len = strlen (str), shift = 0;
+	int len = (int)strlen (str), shift = 0;
```
```suggestion
	size_t len = strlen (str);
	int shift = 0;
```

am11 commented:
@AaronRobinsonMSFT, thanks for your feedback.

The general concern is that changing type on left side or function prototypes entails updating _all_ function calls then callers of those functions. Also, sometimes it may not be possible to make such a change (without more code); e.g. on Windows, `read()`  returns `int` but on Unix, it returns `ssize_t`.

am11 commented on src/mono/mono/component/debugger-agent.c:
```diff
@@ -1398,13 +1398,13 @@ transport_handshake (void)
 
 	do {
-		res = transport_send (handshake_msg, strlen (handshake_msg));
+		res = transport_send (handshake_msg, (int)strlen (handshake_msg));
```
I tried updating the type in prototype and callers, C4267 is then issued for `buffer_init`, `buffer_add_int` and `transport->send`. It is going to be more plumbing for those functions and its callers as well. I don't mind doing it, but do you think we should do it as part of this PR?

AaronRobinsonMSFT replied:
> It is going to be more plumbing for those functions and their callers as well. I don't mind doing it, but do you think we should do it as part of this PR?

Boo. I appreciate your efforts here. I think we need the mono team to make that decision. I prefer to avoid casts when reasonable changes can make them unnecessary - clearly "reasonable" is subjective in this case and something others will need to clarify.

lambdageek replied:
We have to be a little bit careful here.  `transport->send`, in particular is part of an extensibility API that xamarin-macios uses.

I think the cast here is fine.

am11 commented:
@AaronRobinsonMSFT, @lambdageek, the last update is following this pattern:
Before

```c
int len = (int)strlen (foo);
emit_int32(&buf, foo);
```

After
```c
size_t len = strlen (foo);
emit_int32(&buf, (int)foo);
```
i.e. typecast on the line where it is needed.

lateralusX commented on src/mono/mono/component/debugger-agent.c:
```diff
@@ -529,7 +529,7 @@ is_debugger_thread (void)
 		return FALSE;
 
-	return internal->debugger_thread;
+	return (gboolean)internal->debugger_thread;
```
Maybe more readable and portable to do:

return internal->debugger_thread ? TRUE : FALSE

lateralusX commented on src/mono/mono/component/debugger-agent.c:
```diff
@@ -8907,7 +8907,7 @@ thread_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
 			buffer_add_int (buf, 0);
 		} else {
-			const size_t len = strlen (s);
+			const guint32 len = (guint32)strlen (s);
```
Did you intend to have pattern doing type cast on the line used?

am11 replied:
Yup.

lateralusX commented on src/mono/mono/eglib/gstr.c:
```diff
@@ -928,7 +928,11 @@ g_str_from_file_region (int fd, guint64 offset, gsize size)
 	buffer [size] = 0;
 	do {
+#ifdef HOST_WIN32
+		status = _read (fd, buffer, (unsigned)size);
```
unsigned int

am11 replied:
I have consistently used the type (aliases) exactly as reported by the compiler. In this case, it's [this](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/read?view=msvc-170).

lateralusX replied:
There is also g_read in glib.h, we could expand that macro to take parameters and do conversion as part of macro and then use g_read.

lateralusX commented on src/mono/mono/metadata/class-init.c:
```diff
@@ -1739,12 +1739,12 @@ mono_class_interface_match (const uint8_t *bitmap, int id)
-	char *s = (char *)mono_image_alloc (image, len);
+	size_t null_length = strlen ("(null)");
+	size_t len = (s1 ? strlen (s1) : null_length) + (s2 ? strlen (s2) : null_length) + 2;
+	char *s = (char *)mono_image_alloc (image, (int)len);
```
guint

lateralusX commented on src/mono/mono/metadata/class-init.c:
```diff
@@ -1739,12 +1739,12 @@ mono_class_interface_match (const uint8_t *bitmap, int id)
 	int result;
 
-	result = g_snprintf (s, len, "%s%c%s", s1 ? s1 : "(null)", '\0', s2 ? s2 : "(null)");
+	result = g_snprintf (s, (int)len, "%s%c%s", s1 ? s1 : "(null)", '\0', s2 ? s2 : "(null)");
```
g_snprintf uses gulong so maybe we should typecast to that.

lateralusX commented on src/mono/mono/metadata/mempool.c:
```diff
@@ -357,14 +357,13 @@ char*
-	l = strlen (s);
-	res = (char *)mono_mempool_alloc (pool, l + 1);
+	size_t l = strlen (s);
+	res = (char *)mono_mempool_alloc (pool, (int)l + 1);
```
mono_mempool_alloc uses unsigned int.

lateralusX commented on src/mono/mono/metadata/metadata.c:
```diff
@@ -2426,18 +2426,18 @@ static MonoMethodSignature*
+	size_t sigsize, sig_header_size;
 	MonoMethodSignature *ret;
-	sigsize = sig_header_size = MONO_SIZEOF_METHOD_SIGNATURE + sig->param_count * sizeof (MonoType *) + padding;
+	sigsize = sig_header_size = (int)(MONO_SIZEOF_METHOD_SIGNATURE + sig->param_count * sizeof (MonoType *) + padding);
```
sigsize and sig_header_size is already of size_t so no need to downcast to int.

lateralusX commented on src/mono/mono/metadata/metadata.c:
```diff
@@ -2426,18 +2426,18 @@ static MonoMethodSignature*
+	sigsize = sig_header_size = (int)(MONO_SIZEOF_METHOD_SIGNATURE + sig->param_count * sizeof (MonoType *) + padding);
 	if (sig->ret)
-		sigsize += mono_sizeof_type (sig->ret);
+		sigsize += (int)mono_sizeof_type (sig->ret);
```
No need to cast.

lateralusX commented on src/mono/mono/metadata/object.c:
```diff
@@ -2011,12 +2011,12 @@ mono_class_create_runtime_vtable (MonoClass *klass, MonoError *error)
 			imt_table_bytes *= 2;
 		UnlockedIncrement (&mono_stats.imt_number_of_tables);
-		UnlockedAdd (&mono_stats.imt_tables_size, imt_table_bytes);
+		UnlockedAdd (&mono_stats.imt_tables_size, (guint32)imt_table_bytes);
```
UnlockAdd uses gint32.

lateralusX commented on src/mono/mono/metadata/object.c:
```diff
@@ -6018,7 +6018,7 @@ mono_string_new_utf32_checked (const mono_unichar4 *text, gint32 len, MonoError
 	utf16_output = g_ucs4_to_utf16 (text, len, NULL, NULL, NULL);
 
-	gint32 utf16_len = g_utf16_len (utf16_output);
+	gint32 utf16_len = (gint32)g_utf16_len (utf16_output);
```
Switch to size_t and then cast on use?

lateralusX commented on src/mono/mono/metadata/object.c:
```diff
@@ -6217,13 +6217,10 @@ mono_string_new_checked (const char *text, MonoError *error)
-	len = strlen (text);
-
-	ut = g_utf8_to_utf16 (text, len, NULL, &items_written, &eg_error);
+	ut = g_utf8_to_utf16 (text, (int)strlen (text), NULL, &items_written, &eg_error);
```
g_utf8_to_utf16 uses glong.

lateralusX commented on src/mono/mono/metadata/object.c:
```diff
@@ -6928,7 +6925,7 @@ mono_ldstr_metadata_sig (const char* sig, MonoStringHandleOut string_handle, Mon
 	// FIXMEcoop excess handle, use mono_string_new_utf16_checked and string_handle parameter
 
-	MonoStringHandle o = mono_string_new_utf16_handle ((gunichar2*)sig, len, error);
+	MonoStringHandle o = mono_string_new_utf16_handle ((gunichar2*)sig, (guint32)len, error);
```
mono_string_new_utf16_handle uses gint32.

lateralusX commented on src/mono/mono/metadata/sre.c:
```diff
@@ -2052,7 +2052,7 @@ MONO_RESTORE_WARNING
 		str = mono_string_to_utf8_checked_internal ((MonoString*)arg, error);
 		return_if_nok (error);
-		slen = strlen (str);
+		slen = (guint32)strlen (str);
```
size_t slen?

am11 commented:
I think I have addressed the feedback and then some. The existing code is full of inconsistencies, so the feedback is bit subjective. I suggest to leave it to this state (unless, of course, there is a technical issue).

lateralusX commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -1670,7 +1670,7 @@ mono_resolve_patch_target_ext (MonoMemoryManager *mem_manager, MonoMethod *metho
 		char *s;
 
-		len = strlen ((const char *)patch_info->data.target);
+		len = (guint)strlen ((const char *)patch_info->data.target);
```
size_t len and cast on use?

lateralusX commented on src/mono/mono/profiler/aot.c:
```diff
@@ -445,7 +445,7 @@ emit_int32 (MonoProfiler *prof, gint32 value)
 emit_string (MonoProfiler *prof, const char *str)
 {
-	int len = strlen (str);
+	int len = (int)strlen (str);
```
size_t len and cast on use?

lateralusX commented on src/mono/mono/sgen/sgen-marksweep.c:
```diff
@@ -2195,7 +2195,7 @@ major_free_swept_blocks (size_t section_reserve)
 		SGEN_ASSERT (0, num_empty_blocks > 0, "section reserve can't be negative");
 
-		num_empty_blocks_orig = num_empty_blocks;
+		num_empty_blocks_orig = (int)num_empty_blocks;
```
Make num_empty_blocks_orig size_t instead?

lateralusX commented on src/mono/mono/utils/hazard-pointer.c:
```diff
@@ -89,8 +89,8 @@ mono_thread_small_id_alloc (void)
 			g_assert_not_reached ();
-		new_table = mono_bitset_clone (small_id_table, small_id_table->size * 2);
-		id = mono_bitset_find_first_unset (new_table, small_id_table->size - 1);
+		new_table = mono_bitset_clone (small_id_table, (int)small_id_table->size * 2);
```
mono_bitset_clone uses guint32.

lateralusX commented on src/mono/mono/utils/hazard-pointer.c:
```diff
@@ -89,8 +89,8 @@ mono_thread_small_id_alloc (void)
-		new_table = mono_bitset_clone (small_id_table, small_id_table->size * 2);
-		id = mono_bitset_find_first_unset (new_table, small_id_table->size - 1);
+		new_table = mono_bitset_clone (small_id_table, (int)small_id_table->size * 2);
+		id = mono_bitset_find_first_unset (new_table, (int)small_id_table->size - 1);
```
mono_bitset_find_first_unset uses gint.

lateralusX commented on src/mono/mono/utils/monobitset.h:
```diff
@@ -49,18 +49,16 @@ enum {
-	size = tmp_dest->size / MONO_BITSET_BITS_PER_CHUNK; \
-	for (i = 0; i < size; ++i) \
+	size_t size = (tmp_dest->size / MONO_BITSET_BITS_PER_CHUNK); \
+	for (int i = 0; i < size; ++i) \
```
size_t i?

lateralusX commented on src/mono/mono/utils/monobitset.h:
```diff
@@ -49,18 +49,16 @@ enum {
-	size = tmp_dest->size / MONO_BITSET_BITS_PER_CHUNK; \
-	for (i = 0; i < size; ++i) \
+    size_t size = tmp_dest->size / MONO_BITSET_BITS_PER_CHUNK; \
+	for (int i = 0; i < size; ++i) \
```
size_t i?

lateralusX commented:
One thing that came to mind when looking through the changes, since we are mitigating valid warnings using down cast and change of singed/unsigned types, maybe an option could be to cast through a macro or similar and then we could at least make some validation on debug builds that the cast is valid from type x to y? Just a thought since we go over this.

lateralusX commented:
> I think I have addressed the feedback and then some. The existing code is full of inconsistencies, so the feedback is bit subjective. I suggest to leave it to this state (unless, of course, there is a technical issue).

Jupp, understood, I just wrote down the differences that I spotted when looking through changes. If we do an explicit cast I think we should cast to the expected type as much as possible.

lambdageek commented on src/mono/cmake/config.h.in:
```diff
@@ -8,7 +8,6 @@
 #pragma warning(disable:4090) // const problem
 #pragma warning(disable:4146) // unary minus operator applied to unsigned type, result still unsigned
 #pragma warning(disable:4244) // integer conversion, possible loss of data
-#pragma warning(disable:4267) // integer conversion, possible loss of data
```
I wonder how painful it would be to enable `-Wshorten-64-to-32` on clang - it would help catch some of these errors when building on OSX.  Not a lot of the mono team works with the Win32 builds

am11 replied:
1674 hits on M1 (osx-arm64).. project for another weekend. 😅

am11 replied:
On main branch 2002 hits.

lateralusX reviewed (approved):
LGTM, thanks for working through this!

AaronRobinsonMSFT commented:
Failure is https://github.com/dotnet/runtime/issues/66556

SamMonoRT commented:
cc @am11  - if https://github.com/dotnet/runtime/pull/66721 doesn't fix the https://github.com/dotnet/runtime/issues/66718 then we'll be reverting the changes in this PR via https://github.com/dotnet/runtime/pull/66719

akoeplinger commented:
It did fix it so no need to revert :)
